device-code-flow/app.ps1

```diff
@@ -33,19 +33,56 @@ public static class DeviceCodeHelper
 }
 "@ -ReferencedAssemblies $RequiredAssemblies -IgnoreWarnings -WarningAction SilentlyContinue
 
+# Include the MsalCacheHelper used for token caching
+if (-not ('Microsoft.Identity.Client.Extensions.Msal.MsalCacheHelper' -as [type])) {
+    try {
+        Add-Type -Path (Join-Path (Split-Path $PSCommandPath) 'Microsoft.Identity.Client.Extensions.Msal.2.19.6\lib\netcoreapp2.1\Microsoft.Identity.Client.Extensions.Msal.dll')
+    }
+    catch {
+        Write-Error $_
+        return
+    }
+}
 
 # 'Application (client) ID' of app registration in Azure portal - this value is a GUID
 $ClientId = ""
 
 # 'Directory (tenant) ID' of app registration in Azure portal - this value is a GUID
 $TenantId = ""
 
+# Set to 0 by default.  Set this value to 1 to clear the cache.
+$clearCache = 1
+
 # The Device Code flow requires a Public Client Application
 # Build a PublicClientApplication with the $ClientId and $TenantId
-$publicClient = [Microsoft.Identity.Client.PublicClientApplicationBuilder]::Create($ClientId).WithTenantId($TenantId)
+$publicClient = [Microsoft.Identity.Client.PublicClientApplicationBuilder]::Create($ClientId).WithTenantId($TenantId).Build()
+
+# Setup caching.  In this sample a local file named "msal-cache.bin" is where cached tokens will be stored.
+$cacheDir = Split-Path $PSCommandPath
+$storageBuilder = New-Object Microsoft.Identity.Client.Extensions.Msal.StorageCreationPropertiesBuilder("msal-cache.bin", $cacheDir, $ClientId)
+$cacheHelper = [Microsoft.Identity.Client.Extensions.Msal.MsalCacheHelper]::CreateAsync($storageBuilder.Build()).GetAwaiter().GetResult()
+$cacheHelper.RegisterCache($publicClient.UserTokenCache)
```

Comment:
This looks like you're setting up file-based caching.  That's not what we need to show here.  We don't need to cache between invocations, just need to talk about what in-memory cache exists.  None of our other examples (except one, and that was due to an iOS limitation) do we show cross-execution caching.

The caching pattern implementation looks good -- first a call to `AcquireTokenSilent` and then a call to `AcquireTokenWithDeviceCode` if needed.  But I think you might have taken this too far with disk-caching.  HMU if you have questions about this scope creep.

Reply:
@ckittel Thanks for the review!  I had a feeling I went overboard.  I'll review the linked dotnet example and adjust!